### PR TITLE
[JIT] initial support for PEP-585 types

### LINKED ISSUE #56210: JIT: Does not support PEP-585 type annotations
## 🐛 Bug

Type annotations according to PEP-585 do not work with `torch.jit.script`:

```py
import torch

def fun(x: torch.Tensor) -> torch.Tensor:
    xs: list[torch.Tensor] = []
    xs.append(x)
    return xs.pop()

torch.jit.script(fun)

```
```
RuntimeError:
Unknown type constructor list:
  File "~/git/python-tools/test_types.py", line 5
def fun(x: torch.Tensor) -> torch.Tensor:
    xs: list[torch.Tensor] = []
        ~~~~~~~~~~~~~~~~~~ <--- HERE
    xs.append(x)
    return xs.pop()
```

cc @gmagogsfm

### PR DESCRIPTION
**Pull request by walterddr**
Relates to #56210. Initial attempt to make support for `list`, `tuple` and `dict` type in PEP-585.

Test Plan:
- newly added `test_pep585_type`
- CI

### CODE REVIEW

File: torch/csrc/jit/frontend/script_type_parser.cpp [hunk 1]
```diff
@@ -45,7 +45,7 @@ TypePtr ScriptTypeParser::subscriptToType(
       subscript_expr_types.push_back(parseTypeFromExprImpl(expr));
     }
     return TupleType::create(subscript_expr_types);
-  } else if (typeName == "List") {
+  } else if (typeName == "List" || typeName == "list") {
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/script_type_parser.cpp:48]: Thanks for making the change! While we are at it, do you mind adding support for `Tuple` as well?

File: test/jit/test_types.py [hunk 2]
```diff
@@ -21,6 +21,22 @@
                        "instead.")
 
 class TestTypesAndAnnotation(JitTestCase):
+    def test_pep585_type(self):
+        def fn(x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+            xl: list[torch.Tensor] = []
+            xd: dict[str, torch.Tensor] = {}
```
Comment (gmagogsfm) [inline test/jit/test_types.py:27]: Nit:
`List[toch.Tensor]` and `Dict[str, torch.Tensor]` happens to be default types for `[]` and `{}`, so the test case doesn't exact test that PEP 585 is correctly parsed. Do you mind trying some other types like `List[int]`, Dict[str, int]`? Thanks

File: torch/csrc/jit/frontend/script_type_parser.cpp [hunk 3]
```diff
@@ -26,7 +26,7 @@ std::string collectQualname(const Select& select) {
 TypePtr ScriptTypeParser::subscriptToType(
     const std::string& typeName,
     const Subscript& subscript) const {
-  if (typeName == "Tuple") {
+  if (typeName == "Tuple" || typeName == "tuple") {
```
Comment (walterddr) [inline torch/csrc/jit/frontend/script_type_parser.cpp:29]: Added tuple checks

**Review by walterddr — COMMENTED** [review]
addressed all comments @gmagogsfm could you take another look.

File: test/jit/test_types.py [hunk 4]
```diff
@@ -21,6 +21,22 @@
                        "instead.")
 
 class TestTypesAndAnnotation(JitTestCase):
+    def test_pep585_type(self):
+        def fn(x: torch.Tensor) -> Tuple[Tuple[torch.Tensor], Dict[str, int]]:
```
[the remark this answers is not in the document]
Comment (walterddr) [reply test/jit/test_types.py]: actually I just checked and this doesn't work --> return type PEP585 requires py3.9 and we are still minimum supporting py3.6. reverting this

[the remark this answers is not in the document]
Comment (gmagogsfm) [reply test/jit/test_types.py:25]: I see, thanks for digging into this. IMHO we can limit the test to only run on Py 3.9+. It is still valuable to use PEP585 style for return type for more comprehensive coverage.

[the remark this answers is not in the document]
Comment (walterddr) [reply test/jit/test_types.py:25]: -_-!!!
i dont think we have any CI pipeline that's currently running py3.9.

Comment (walterddr) [inline test/jit/test_types.py:25]:
```suggestion
        # TODO add test to use PEP585 type annotation for return type after py3.9
        # see: https://www.python.org/dev/peps/pep-0585/#id5
        def fn(x: torch.Tensor) -> Tuple[Tuple[torch.Tensor], Dict[str, int]]:
```

[the remark this answers is not in the document]
Comment (gmagogsfm) [reply test/jit/test_types.py:27]: Ouch. A TODO sounds good then.

**Review by gmagogsfm — APPROVED** [review]

**Comment by randolf-scholz** [comment]
Hey, since python 3.6 is EOL now, is this going to make it into pytorch 1.11 ? https://github.com/pytorch/pytorch/milestone/26

**Pull request closed without merging**